File: VB/Security_AllowLinkUnlink.Win/WinApplication.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports DevExpress.ExpressApp.Win
Imports DevExpress.ExpressApp.Security

Partial Public Class Security_AllowLinkUnlinkWindowsFormsApplication
	Inherits WinApplication
	Public Sub New()
		InitializeComponent()
	End Sub

	Private Sub Security_AllowLinkUnlinkWindowsFormsApplication_DatabaseVersionMismatch(ByVal sender As Object, ByVal e As DevExpress.ExpressApp.DatabaseVersionMismatchEventArgs) Handles MyBase.DatabaseVersionMismatch
#If EASYTEST Then
        e.Updater.Update()
        e.Handled = True
#Else
        If System.Diagnostics.Debugger.IsAttached Then
            e.Updater.Update()
            e.Handled = True
        Else
            Throw New InvalidOperationException( _
             "The application cannot connect to the specified database, because the latter doesn't exist or its version is older than that of the application." & Constants.vbCrLf & _
             "The automatic update is disabled, because the application was started without debugging." & Constants.vbCrLf & _
             "You should start the application under Visual Studio, or modify the " & _
             "source code of the 'DatabaseVersionMismatch' event handler to enable automatic database update, " & _
             "or manually create a database using the 'DBUpdater' tool.")
        End If
#End If
    End Sub

    Private Sub Security_AllowLinkUnlinkWindowsFormsApplication_LastLogonParametersRead(ByVal sender As System.Object, ByVal e As DevExpress.ExpressApp.LastLogonParametersReadEventArgs) Handles MyBase.LastLogonParametersRead
        Dim logonParams As AuthenticationStandardLogonParameters = CType(e.LogonObject, AuthenticationStandardLogonParameters)
        If String.IsNullOrEmpty(logonParams.UserName) Then
            logonParams.UserName = "Sam"
        End If
    End Sub
End Class

File: CS/Security_AllowLinkUnlink.Win/WinApplication.cs
using DevExpress.ExpressApp.Xpo;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using DevExpress.ExpressApp.Win;
using DevExpress.ExpressApp.Updating;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Security;

namespace Security_AllowLinkUnlink.Win {
    public partial class Security_AllowLinkUnlinkWindowsFormsApplication : WinApplication {
        protected override void CreateDefaultObjectSpaceProvider(CreateCustomObjectSpaceProviderEventArgs args) {
            args.ObjectSpaceProvider = new XPObjectSpaceProvider(args.ConnectionString, args.Connection);
        }
        public Security_AllowLinkUnlinkWindowsFormsApplication() {
            InitializeComponent();
        }

        private void Security_AllowLinkUnlinkWindowsFormsApplication_DatabaseVersionMismatch(object sender, DevExpress.ExpressApp.DatabaseVersionMismatchEventArgs e) {
#if EASYTEST
			e.Updater.Update();
			e.Handled = true;
#else
            if(System.Diagnostics.Debugger.IsAttached) {
                e.Updater.Update();
                e.Handled = true;
            }
            else {
                throw new InvalidOperationException(
                    "The application cannot connect to the specified database, because the latter doesn't exist or its version is older than that of the application.\r\n" +
                    "The automatic update is disabled, because the application was started without debugging.\r\n" +
                    "You should start the application under Visual Studio, or modify the " +
                    "source code of the 'DatabaseVersionMismatch' event handler to enable automatic database update, " +
                    "or manually create a database using the 'DBUpdater' tool.");
            }
#endif
        }

        private void Security_AllowLinkUnlinkWindowsFormsApplication_LastLogonParametersRead(object sender, LastLogonParametersReadEventArgs e) {
            AuthenticationStandardLogonParameters logonParams = (AuthenticationStandardLogonParameters)e.LogonObject;
            if(string.IsNullOrEmpty(logonParams.UserName)) {
                logonParams.UserName = "Sam";
            }
        }
    }
}
